File: jupyterqt/models/notebook_model.py
import uuid
from dataclasses import dataclass, field

from jupyterqt.models.cell_model import CellModel, CellType


@dataclass
class NotebookModel:
    notebookId: str
    path: str
    cells: list[CellModel]
    kernel_id: str | None = None
    kernel_name: str = "python3"
    metadata: dict = field(default_factory=dict)
    nbformat: int = 4
    nbformat_minor: int = 5

    @staticmethod
    def fromIpynbDict(path: str, data: dict) -> "NotebookModel":
        cells = [CellModel.fromIpynbCell(c) for c in data.get("cells", [])]
        if not cells:
            cells = [CellModel.new(CellType.CODE)]
        metadata = data.get("metadata", {})
        kernel_name = metadata.get("kernelspec", {}).get("name", "python3")
        return NotebookModel(
            notebookId=str(uuid.uuid4()),
            path=path,
            cells=cells,
            metadata=metadata,
            kernel_name=kernel_name,
            nbformat=data.get("nbformat", 4),
            nbformat_minor=data.get("nbformat_minor", 5),
        )

    def toIpynbDict(self) -> dict:
        return {
            "nbformat": self.nbformat,
            "nbformat_minor": self.nbformat_minor,
            "metadata": self.metadata,
            "cells": [c.toIpynbCell() for c in self.cells],
        }

    def getCell(self, cellId: str) -> CellModel | None:
        for c in self.cells:
            if c.cellId == cellId:
                return c
        return None

    def addCell(self, cell_type: CellType = CellType.CODE,
                 index: int | None = None) -> CellModel:
        cell = CellModel.new(cell_type)
        if index is None:
            self.cells.append(cell)
        else:
            self.cells.insert(index, cell)
        return cell

    def removeCell(self, cellId: str) -> None:
        self.cells = [c for c in self.cells if c.cellId != cellId]

    def moveCell(self, cellId: str, new_index: int) -> None:
        cell = self.getCell(cellId)
        if cell is None:
            return
        self.cells.remove(cell)
        self.cells.insert(new_index, cell)

    def indexOf(self, cellId: str) -> int:
        for i, c in enumerate(self.cells):
            if c.cellId == cellId:
                return i
        return -1

File: cpp/src/models/NotebookModel.cpp
#include "NotebookModel.h"
#include <QJsonArray>
#include <QFileInfo>

NotebookModel::NotebookModel(QObject* parent)
    : QObject(parent)
{
}

NotebookModel::~NotebookModel() {
    qDeleteAll(m_cells);
}

QString NotebookModel::name() const {
    return QFileInfo(m_path).fileName();
}

CellModel* NotebookModel::cell(const QString& cellId) const {
    for (auto* c : m_cells) {
        if (c->cellId() == cellId) return c;
    }
    return nullptr;
}

int NotebookModel::cellIndex(const QString& cellId) const {
    for (int i = 0; i < m_cells.size(); ++i) {
        if (m_cells[i]->cellId() == cellId) return i;
    }
    return -1;
}

void NotebookModel::loadFromJson(const QJsonObject& nbJson) {
    qDeleteAll(m_cells);
    m_cells.clear();

    m_metadata = nbJson[QStringLiteral("metadata")].toObject();
    m_nbformat = nbJson[QStringLiteral("nbformat")].toInt(4);
    m_nbformat_minor = nbJson[QStringLiteral("nbformat_minor")].toInt(5);

    auto kernelspec = m_metadata[QStringLiteral("kernelspec")].toObject();
    m_kernelName = kernelspec[QStringLiteral("name")].toString(QStringLiteral("python3"));

    auto cellsArr = nbJson[QStringLiteral("cells")].toArray();
    for (const auto& cv : cellsArr) {
        auto* cell = CellModel::fromJson(cv.toObject(), this);
        m_cells.append(cell);
    }
}

QJsonObject NotebookModel::toJson() const {
    QJsonObject nb;
    nb[QStringLiteral("nbformat")] = m_nbformat;
    nb[QStringLiteral("nbformat_minor")] = m_nbformat_minor;
    nb[QStringLiteral("metadata")] = m_metadata;

    QJsonArray cells;
    for (auto* c : m_cells) {
        cells.append(c->toJson());
    }
    nb[QStringLiteral("cells")] = cells;
    return nb;
}

CellModel* NotebookModel::addCell(CellType type, int index) {
    auto* cell = new CellModel(this);
    cell->setCellType(type);

    if (index < 0 || index >= m_cells.size()) {
        m_cells.append(cell);
        emit cellAdded(m_cells.size() - 1, cell);
    } else {
        m_cells.insert(index, cell);
        emit cellAdded(index, cell);
    }
    return cell;
}

void NotebookModel::removeCell(const QString& cellId) {
    int idx = cellIndex(cellId);
    if (idx < 0) return;
    auto* c = m_cells.takeAt(idx);
    emit cellRemoved(cellId);
    c->deleteLater();
}

void NotebookModel::moveCell(const QString& cellId, int newIndex) {
    int idx = cellIndex(cellId);
    if (idx < 0) return;
    newIndex = qBound(0, newIndex, m_cells.size() - 1);
    if (idx == newIndex) return;
    auto* c = m_cells.takeAt(idx);
    m_cells.insert(newIndex, c);
    emit cellMoved(cellId, newIndex);
}
